Simbody/tests/TestConstraints.cpp

```diff
@@ -582,6 +582,152 @@ void testConstraintMatrices() {
+    MobilizedBody& first = matter.updMobilizedBody(MobilizedBodyIndex(1));
+    MobilizedBody& second = matter.updMobilizedBody(MobilizedBodyIndex(2));
+    MobilizedBody& fifth = matter.updMobilizedBody(MobilizedBodyIndex(5));
+    MobilizedBody& last = matter.updMobilizedBody(MobilizedBodyIndex(NUM_BODIES));
+
+    // Add a body on a free joint to body 5 then weld it.
+    MobilizedBody::Free extra(fifth, Transform(),
+        MassProperties(1, Vec3(.01,.02,.03), Inertia(1,1.1,1.2)), Transform());
+    Constraint::Weld weld(extra, fifth);
+
+    Constraint::Ball ball(first, last);
+    Constraint::ConstantAcceleration accel2(second, MobilizerUIndex(1), .01);
+    Constraint::ConstantSpeed speed5(fifth, MobilizerUIndex(1), .1);
+
+    // Obtain constraint errors from realizing to Acceleration.
+    Vector udotFromForward, pvaerrFromForward;
+    State state;
+    {
+        State stateForward;
+        createState(system, stateForward); // Realizes to Acceleration.
+        udotFromForward = stateForward.getUDot();
+        pvaerrFromForward = stateForward.getUDotErr();
+        state = stateForward; // This copy assignment invalidates the cache.
+    }
+
+    const int nu = matter.getNU(state);
+    const int m  = matter.getNUDotErr(state);
+
+    // Ensure that we get an exception if not realized to Velocity yet.
+    {
+        system.realize(state, SimTK::Stage::Position);
+        SimTK_TEST(state.getSystemStage() == SimTK::Stage::Position);
+        Vector output; Vector udot(nu);
+        SimTK_TEST_MUST_THROW_EXC(
+                matter.calcConstraintAccelerationErrors(state, udot, output),
+                SimTK::Exception::ErrorCheck);
+    }
+
+    // Obtain constraint errors using the operator.
+    Vector biasFromOperator; // used in a subsequent test.
+    {
+        State stateOperator = state;
+        // Realizing to Velocity should be sufficient.
+        system.realize(stateOperator, SimTK::Stage::Velocity);
+
+        // With udot=0, errors should be nonzero.
+        Vector outputWithZeroUDot;
+        matter.calcConstraintAccelerationErrors
+                (stateOperator, Vector(nu, 0.0), outputWithZeroUDot);
+        SimTK_TEST(outputWithZeroUDot.norm() > 1e-6);
+        biasFromOperator = outputWithZeroUDot; // used in a subsequent test.
+
+        // Ensure the operator did not internally realize to a later stage.
+        SimTK_TEST(stateOperator.getSystemStage() == SimTK::Stage::Velocity);
+
+        // With udot from forward dynamics, errors should be close to zero.
+        Vector pvaerrFromOperator;
+        matter.calcConstraintAccelerationErrors
+                (stateOperator, udotFromForward, pvaerrFromOperator);
+
+        SimTK_TEST(pvaerrFromOperator.norm() < 1e-10);
+
+        // The same code should have been executed when using forward
+        // dynamics or the operator, so the errors should be exactly the same.
+        SimTK_TEST(pvaerrFromOperator.size() == m);
```

Comment:
This condition is probably too strict. It would be OK if later someone didn't run *exactly* the same code. I would suggest checking to SignificantReal (1e-14 ish) which is what you get with `SimTK_TEST_EQ`.